### PR TITLE
coord: manage Postgres source external state

### PR DESCRIPTION
**Pull request by JLDLaughlin**
This PR does a few things:
- Introduces the scaffolding for the new `ExternalState` enum that was settled upon in our [recent design review and discussion](https://github.com/MaterializeInc/materialize/pull/6655).
- Correctly drops an upstream Postgres replication slot when a Postgres source is dropped using the new `ExternalState` enum.
- Adds a new `postgres-verify-slot` action to testdrive with tests to confirm that we are correctly dropping the upstream replication slots.

This implementation works as long as we enforce that users can only create materialized Postgres sources, but will need to be tweaked when we allow unmaterialized Postgres sources. This future work is captured in [this refactoring issue](https://github.com/MaterializeInc/materialize/issues/6723).

### CODE REVIEW

File: src/postgres-util/src/lib.rs [hunk 1]
```diff
@@ -134,3 +134,34 @@ pub async fn publication_info(
...
+    let (client, connection) = tokio_postgres::connect(&conn, NoTls).await?;
+    tokio::spawn(connection);
+
+    for slot in slots {
+        let active_pid = query_pg_replication_slots(&client, slot).await?;
+        if let Some(pid) = active_pid {
+            client
+                .query("SELECT pg_terminate_backend($1)", &[&pid])
+                .await?;
+        }
```
Comment (benesch) [inline src/postgres-util/src/lib.rs:148]: Oh, shoot. I suspect calling `pg_terminate_backend` will be problematic in a lot of real world deployments. That requires a superuser account, and you probably don't want to give Materialize that kind of access to your Postgres instance. Ugh. We'll probably need to sort out some of way of forcing the worker threads to drop their connections before `drop_replication_slots` runs.

**Review by benesch — COMMENTED** [review]
So I think this framework broadly makes sense, but I am nervous about making changes to our on-disk state without having the actual _need_ for that extra on-disk state wired up. More concretely, I think wiring up external-state per source instances might require the ability to have multiple `(global_id, state)` pairs per source instance, and that will require a slightly different API and, perhaps more importantly, a slightly different looking table on disk.

What do you think about basically hacking in an `if` statement to `catalog_transact` for now that checks whether a Postgres source is dropped and, if so, makes the call to `drop_replication_slots`?

File: src/postgres-util/src/lib.rs (hunk 1 above)
Comment (petrosagg) [reply src/postgres-util/src/lib.rs:148]: If we drop the replication slot through a replication connection using the `DROP_REPLICATION_SLOT` command we can use the `WAIT` modifier there that will wait for all users to stop.

That plus a reasonable timeout should do the trick with our existing permissions, how does it sound?

**Review by petrosagg — APPROVED** [review]
This looks good to me modulo Nikhil's comments

**Comment by petrosagg** [comment]
@benesch we will definitely need that. Here is the full plan we discussed with Jessica:

> We create two tracking tables, a) `mz_source_instance` that has an `gid` and a `state`  b) `mz_source_instance_index` that has a `source_instance_id` and an `index_id` that describes the many to many relationship between the indexes and source instances.
> The rule is that every index using a source must have a corresponding entry in `mz_source_instance_index` that links it with a source instance. Then whenever we create `DataflowDesc` structs that have multiple entries in `export_index` we can express that in the tracking tables by sharing a single `mz_source_instance` entry and this will also allow us to re-create the same topology on restart

The hope was to do the many to many part later but it's true that we'll then need to write migration code. Doing your hack while implementing the full proposal sounds like a good middle ground to me.

**Comment by JLDLaughlin** [comment]
Closing in favor of #6785, although we will want to do something like this fairly soon!

**Pull request closed without merging**